include/mockturtle/utils/recursive_cost_functions.hpp

```diff
@@ -0,0 +1,111 @@
+/* mockturtle: C++ logic network library
+ * Copyright (C) 2018-2021  EPFL
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following
+ * conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/*!
+  \file recursive_cost_functions.hpp
+  \brief Various recursive cost functions for (optimization) algorithms
+
+  \author Hanyu Wang
+*/
+
+#pragma once
+
+#include <cstdint>
+
+#include "../traits.hpp"
+
+namespace mockturtle
+{
+
+/*! \brief (Recursive) customizable cost function
+ * 
+ * This is the base type of a customizable cost function. Two operators need to be defined.
+ */
```

Comment:
Since this is something the users have to define for their applications, they need more information to learn how it works. You need to explain, at least, what is "context". A link or reference to your paper will also help (however, the IWLS proceeding is in principle not public); otherwise, we need the formulas and explanations here as clear as in the paper. You can also mention in the documentation that there are examples defined in this file.